File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int best = INT_MIN;
    int dfs(TreeNode* root){
        if (!root)
            return -1;
        int lh = dfs(root->left);
        int rh = dfs(root->right);
        best = max(best,lh+rh+2);
        return max(lh,rh)+1;
    }
    int diameterOfBinaryTree(TreeNode* root) {
        dfs(root);
        return best;
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def dfs(root,best):
            if not root:
                return (-1,best)
            if not root.left and not root.right:
                return (0,best)
            lh,leftBest = dfs(root.left,best)
            rh,subTreesBest = dfs(root.right,leftBest)
            best = max(subTreesBest,lh+rh+2)
            return (max(lh,rh)+1,best)
        _,best = dfs(root,0)
        return best